### PR TITLE
Remove assignability cases in getNarrowedType + an isArray improvement for readonly arrays 

### LINKED ISSUE #17002: Array.isArray type narrows to any[] for ReadonlyArray<T>
**TypeScript Version:**  2.4.1

**Code**

```ts
declare interface T {
  foo: string;
}

const immutable: T | ReadonlyArray<T> = [];

if (Array.isArray(immutable)) {
  const x = immutable; // Any[]  - Should be ReadonlyArray<T>
} 

const mutable: T | Array<T> = [];

if (Array.isArray(mutable)) {
  const x = mutable; // T[]
} 
```

**Expected behavior:** Should type narrow to ```ReadonlyArray<T>```, or at the very least ```T[]```.

**Actual behavior:** Narrows to ```any[]```.  Doesn't trigger warnings in noImplicitAny mode either.

### LINKED ISSUE #31155: 'instanceof' changes type outside of 'if' statement
**TypeScript Version:**  3.4.*

**Search Terms:**

instanceof

**Code**

```ts
// @strictNullChecks: true
interface OnChanges {
    onChanges(changes: Record<string, unknown>): void
}
interface Validator {
    validate(): null | Record<string, unknown>;
}

class C {
    validate() {
        return {}
    }
}
function foo() {
    let v: Validator & Partial<OnChanges> = null as any;
    if (v instanceof C) {
    }
    if (v.onChanges) { // error here
        v.onChanges({});
    }
}
```

**Expected behavior:**

No error.

**Actual behavior:**

```
Property 'onChanges' does not exist on type 'C | (Validator & Partial<OnChanges>)'.
  Property 'onChanges' does not exist on type 'C'.
````

`instanceof` shouldn't change the type of the variable outside of the `if` statement.

This only happens with `strictNullChecks` enabled.

**Playground Link:** https://typescript-play.js.org/#code/JYOwLgpgTgZghgYwgAgPIgMIAs4gOYQDOyA3gFDKXID2mO+RAFAvQYQFzIBKEC1UAEwA8hMFFB4ANMgCuIANYhqAdxAA+AJScAbtWACyAXzKhIsRCgBqcADb64YfqQpVtt+5EZbkIGTZvIAD7cvPzCouL40nKKKuoA3EZkZAg2cITEGM5UyG52Ag4QXtk5VFAQYDJQIKTGOcbGMHIIYMC0yDDU1MXkOTYVuZzW+Q5OAGTIAApwUK22QujYuGxqyAC8Pn4B6ci4AJ6JOcAwyIzayKCiuEjUJxgaJZR1VMen2gB0tEsMhA+9pZQPl9WEwSIYNIcqA0gA

**Related Issues:**

### PR DESCRIPTION
**Pull request by orta**
Fixes #31155 
Fixes #17002

There are two pieces to this PR:

 - The fix for #31155 which came from #32443
 - Unblocking #17002 which stopped #32443

The 2nd was blocked on :

> TLDR: any[] is assignable to IAction | ReadonlyArray<IAction>, but it's not a subtype. The issue here is that readonly arrays are not subtypes of mutable arrays so we end up falling back to the intersection case.

Where :

```ts
interface IAction {
    x: string;
}

declare const arg: IAction | ReadonlyArray<IAction>;
const x = Array.isArray(arg) ? arg : [arg];
```

Is now legal due to changes in the `isArray` definition which takes into account `readonly` arrays.

/cc @jack-williams

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this
@typescript-bot user test this

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this
@typescript-bot user test this

**Comment by jack-williams** [comment]
Seems to be tripping up when the guarded value has type `any` and it picks both parts of the conditional when it should really pick the false branch. Could something like the following work?
```ts
declare function isArray<T>(arg: T | {}): arg is T extends readonly any[] ? ReadonlyFromCoercion<T>: any[];
type ReadonlyFromCoercion<T extends readonly unknown[]> = 0 extends (1 & T) ? any[]: T;
```

**Comment by weswigham** [comment]
```ts
isArray<T>(arg: T | {}): arg is T extends readonly any[] ? (unknown extends T ? never : readonly any[]) : any[];
```
should also work to filter out `any` from the `true` branch.

**Comment by orta** [comment]
@typescript-bot test this
@typescript-bot user test this

**Comment by orta** [comment]
OK, so that does clean up _most_ of the errors in the RWC in the azure side - it looks like just a few rxjs issues and they all stem to their own custom [`isArray`](https://github.com/ReactiveX/rxjs/blob/master/src/internal/util/isArray.ts#L1) function which looks like:

```ts
export const isArray = (() => Array.isArray || (<T>(x: any): x is T[] => x && typeof x.length === 'number'))();
```

We're still seeing issues in this PR: https://github.com/typescript-bot/TypeScript/pull/53

```
[XX:XX:XX] Error: /vscode/src/vs/base/browser/ui/actionbar/actionbar.ts(628,9): Type '(readonly IAction[] & any[]) | (IAction & any[]) | (IAction | readonly IAction[])[]' is not assignable to type 'readonly IAction[]'.
   Type '(IAction | readonly IAction[])[]' is not assignable to type 'readonly IAction[]'.
     Type 'IAction | readonly IAction[]' is not assignable to type 'IAction'.
       Type 'readonly IAction[]' is missing the following properties from type 'IAction': id, label, tooltip, class, and 4 more
```

this could be because they have their [own `isArray`](https://github.com/microsoft/vscode/blob/75e134ae1d398b7d8c971c0cbfdc2e56c319a20a/src/vs/base/common/types.ts#L11) in vscode which wouldn't propagate the `readonly`:

```ts
export function isArray(array: any): array is any[] {
	return Array.isArray(array);
}
```

I could update the vscode definition to the new definition to see if that fixes it

**Comment by weswigham** [comment]
Yeah, the custom `isArray` functions would need to adopt the same type, both to be assignment compatible and to behave the same way.

**Comment by orta** [comment]
Looks like I need a build to verify rxjs 

@typescript-bot pack this

**Comment by orta** [comment]
@typescript-bot pack this

**Comment by orta** [comment]
I'm not getting fails in rxjs with this build of TS, I think this is good for one last look and can be merged if we're happy.

**Comment by orta** [comment]
Ping on reviews @sandersn / @sheetalkamat - as it's a potential breakage, I'd like to get it running in nightlies earlier in 4.1

**Review by sandersn — APPROVED** [review]
I'm OK with this if you are prepared to watch closely for breaks and revert it before 4.1 RC if it's too bad.

It seems like a good change, it's just that any change to the lib is guaranteed breaky.

**Comment by orta** [comment]
I'll merge on tuesday when I'm off DT rotation 👍🏻

**Comment by orta** [comment]
Today is that tuesday, I'm going to merge 👍🏻

**Comment by RyanCavanaugh** [comment]
Example for blog post
```ts
declare const p: string | ReadonlyArray<string>;
if (Array.isArray(p)) {
  // 4.0: p: any[]
  // 4.1: p: readonly string[];

  // 4.0: OK
  // 4.1: Error
  p.push(0);
}
```

**Comment by laughinghan** [comment]
FWIW, possibly a better example for the blogpost is one where TypeScript narrowed the type incorrectly and unsafely, rather than merely providing a weaker-than-desired type, e.g.:

```ts
declare var foo: string[] | readonly number[] | null

if (foo instanceof Array) {
    foo; // at runtime, this may be string[] or readonly number[],
         // however, TypeScript <4.1 would narrow to just string[]

    foo.push('str'); // incorrect and unsafe, but allowed by <4.1, now errors in Nightly
} else {
    foo; // at runtime, strictly null, however TypeScript <4.1 would
         // narrow to readonly number[] | null
}
```

[(Playground Link)](https://www.typescriptlang.org/play?ts=4.0.5#code/CYUwxgNghgTiAEA3W8BmB7dAueBnALjAJYB2A5gNoC68APvHFMOiRAJ7wkCuAtgEYgY1Opy4QIAKAlFU8ABQZ08UgSgkwIdLICCMGFDYBKeAG8J8C2kwBueAHo78KPgZcS+IjxAAaePgAWRLjwPAbwAniEpJQ06DAMIEws7KL8gtTe5pbZ9o7+6ADuIIiCvgAqbAAOIADKYMSVLgA8ACwAdACM8AXoYsCcsDCFfkoAVlwEkcTk1FLZim2VE-5yAOQEMKuGtg7K6nFwYC5q-W64UKg+4VzH4oUg-XwcrZ2+JMOCQzDBpPAAckQyP58OwJABfeAgCC4BBmeY2XJOFwwNweLy+DZEI4pbjiXz5IoleIVap1BrNdpdHp9LI5Cy7EiDYb4JSMZisDjcNJCGj0XGSMFAA)

**Comment by felixfbecker** [comment]
@orta I think this broke passing `Array.isArray` to other type-guard-accepting functions, e.g. `array.filter()`: https://github.com/microsoft/TypeScript/issues/41610

**Comment by victorgarciaesgi** [comment]
> @orta I think this broke passing `Array.isArray` to other type-guard-accepting functions, e.g. `array.filter()`: #41610

It brokes a lot de `array.map` for me too. Only happens on recursive types for me #41609

**Comment by MartinJohns** [comment]
@orta This PR has the `Breaking Change` label, but I don't see any mention of it on the "Breaking Changes" documentation:

https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#typescript-41

**Comment by yseymour** [comment]
This is broken for generics and other nontrivial types:
```ts
declare function f1<T extends any[]>(array: T): void;
function f2<T>(thing: T) {
    if (Array.isArray(thing)) {
        f1(thing);  // Argument of type 'T & (T extends readonly any[] ? unknown extends T ? never : readonly any[] : any[])' is not assignable to parameter of type 'any[]'.
    }
}
```
In 4.0, the type of `thing` was narrowed to `T & any[]` which was OK.

**Comment by MartinJohns** [comment]
Another issue caused by the improvement: #41714
